### PR TITLE
[BEAM-9383] Staging Dataflow artifacts from environment

### PR DESCRIPTION
**Pull request by ihji**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ihji** [comment]
This PR depends on #10621. needs to be rebased before merging.

**Comment by ihji** [comment]
R: @robertwb, @chamikaramj

**Comment by chamikaramj** [comment]
Retest this please

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 1]
```diff
@@ -752,6 +759,27 @@ private Debuggee registerDebuggee(CloudDebugger debuggerClient, String uniquifie
     }
   }
 
+  private List<DataflowPackage> stageArtifacts(RunnerApi.Pipeline pipeline) {
+    ImmutableList.Builder<String> filesToStageBuilder = ImmutableList.builder();
+    for (Map.Entry<String, RunnerApi.Environment> entry :
+        pipeline.getComponents().getEnvironmentsMap().entrySet()) {
+      for (RunnerApi.ArtifactInformation info : entry.getValue().getDependenciesList()) {
+        if (!BeamUrns.getUrn(RunnerApi.StandardArtifacts.Types.FILE).equals(info.getTypeUrn())) {
```
Comment (chamikaramj) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: Check the ROLE as well ?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 2]
```diff
@@ -752,6 +759,27 @@ private Debuggee registerDebuggee(CloudDebugger debuggerClient, String uniquifie
     }
   }
 
+  private List<DataflowPackage> stageArtifacts(RunnerApi.Pipeline pipeline) {
```
Comment (chamikaramj) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: Please add a unit test.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 3]
```diff
@@ -563,19 +562,31 @@ def _gcs_file_copy(self, from_path, to_path):
     with open(from_path, 'rb') as f:
       self.stage_file(to_folder, to_name, f, total_size=total_size)
 
-  def _stage_resources(self, options):
+  def _stage_resources(self, pipeline, options):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Please add a unit test

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 4]
```diff
@@ -563,19 +562,31 @@ def _gcs_file_copy(self, from_path, to_path):
...
+    resources = []
+    for _, env in pipeline.components.environments.items():
+      for dep in env.dependencies:
+        if dep.type_urn != common_urns.artifact_types.FILE.urn:
+          raise RuntimeError('unsupported artifact type %s' % dep.type_urn)
+        if dep.role_urn != common_urns.artifact_roles.STAGING_TO.urn:
+          raise RuntimeError('unsupported role type %s' % dep.role_urn)
+        type_payload = beam_runner_api_pb2.ArtifactFilePayload.FromString(
+            dep.type_payload)
+        role_payload = \
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Nit: Pls use ( instead of \ for formatting.

Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: done

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 3 above)
Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 2 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: It's private method so we can't directly test it. Existing test (such as https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java#L758) should cover this method too.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 1 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: Currently, DataflowRunner auto-generates staging names based on source file names. So ROLE (STAGING_TO) is ignored anyway. We could change this behavior later in BEAM-9455.

**Comment by ihji** [comment]
@chamikaramj PTAL and rerun the tests.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Run Java PreCommi

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by chamikaramj** [comment]
Run Portable_Python PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Portable_Python PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Looks like this needs a rebase.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by ihji** [comment]
@chamikaramj @robertwb : PTAL. Thanks!

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 5]
```diff
@@ -772,6 +783,88 @@ private Debuggee registerDebuggee(CloudDebugger debuggerClient, String uniquifie
...
+        options.as(DataflowPipelineDebugOptions.class).getOverrideWindmillBinary();
+    String dataflowWorkerJar = options.getDataflowWorkerJar();
+    if (dataflowWorkerJar != null && !dataflowWorkerJar.isEmpty()) {
+      // Put the user specified worker jar at the start of the classpath, to be consistent with the
+      // built in worker order.
+      pathsToStageBuilder.add(dataflowWorkerJar);
+      aliasMapBuilder.put(dataflowWorkerJar, "dataflow-worker.jar");
+    }
+    for (String path : options.getFilesToStage()) {
+      if (path.contains("=")) {
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:836]: Is this = support a dataflow-only thing? Seems we don't support that in `Environments.getArtifacts()` (but if we did nearly all of this code could go away).

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 6]
```diff
@@ -336,25 +323,26 @@ public DataflowPackage stageToFile(
...
-        String[] components = classpathElement.split("=", 2);
-        sourcePackage.setName(components[0]);
-        sourcePackage.setLocation(components[1]);
-      } else {
-        sourcePackage.setName(null);
-        sourcePackage.setLocation(classpathElement);
+    for (StagedFile classpathElement : classpathElements) {
+      DataflowPackage targetPackage = classpathElement.getStagedPackage();
+      String source = classpathElement.getSource();
+      if (source.contains("=")) {
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:329]: Why do we have to handle this here and above?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 7]
```diff
@@ -442,45 +448,56 @@ public static StagingResult uploaded(PackageAttributes attributes) {
...
   @AutoValue
   abstract static class PackageAttributes {
-
-    public static PackageAttributes forFileToStage(File source, String stagingPath)
+    public static PackageAttributes forFileToStage(File file, String stagingPath)
         throws IOException {
+      return forFileToStage(file.getPath(), null, stagingPath);
+    }
 
+    public static PackageAttributes forFileToStage(
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:456]: This seems highly redundant with what we're already doing in Environments.getArtifacts. Can't we ensure we have a set of (existing, non-directory) files in the environment, and then have these utilities simply do the uploading?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 8]
```diff
@@ -462,7 +462,8 @@ def run_pipeline(self, pipeline, options):
     use_fnapi = apiclient._use_fnapi(options)
     from apache_beam.transforms import environments
     default_environment = environments.DockerEnvironment.from_container_image(
-        apiclient.get_container_image_from_options(options))
+        apiclient.get_container_image_from_options(options),
+        artifacts=environments.python_sdk_dependencies(options))
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:466]: Should we push populating artifacts into from_container_image?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py [hunk 9]
```diff
@@ -102,7 +102,8 @@ def setUp(self):
         '--staging_location=ignored',
         '--temp_location=/dev/null',
         '--no_auth',
-        '--dry_run=True'
+        '--dry_run=True',
+        '--sdk_location=container'
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py:106]: Why this change?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java [hunk 10]
```diff
@@ -210,56 +209,55 @@ public static Environment createProcessEnvironment(
     }
   }
 
-  private static List<ArtifactInformation> getArtifacts(List<String> stagingFiles) {
-    Set<String> pathsToStage = Sets.newHashSet(stagingFiles);
+  public static List<ArtifactInformation> getArtifacts(
+      List<String> stagingFiles, StagingFileNameGenerator generator) {
     ImmutableList.Builder<ArtifactInformation> artifactsBuilder = ImmutableList.builder();
-    for (String path : pathsToStage) {
+    for (String path : ImmutableSet.copyOf(stagingFiles)) {
```
[the remark this answers is not in the document]
Comment (ihji) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java:215]: `ImmutableSet` preserves the order but I think we don't need to make a copy here. Will use `LinkedHashSet` instead.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 11]
```diff
@@ -784,7 +877,25 @@ public DataflowPipelineJob run(Pipeline pipeline) {
...
+
+    sdkComponents.registerEnvironment(
+        defaultEnvironmentForDataflow
+            .toBuilder()
+            .addAllDependencies(getDefaultArtifacts())
+            .build());
+
+    RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(pipeline, sdkComponents, true);
+
+    LOG.debug("Portable pipeline proto:\n{}", TextFormat.printToString(pipelineProto));
```
[the remark this answers is not in the document]
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:896]: This debug log is not new. It's just relocated. Do you think it would be better to remove this?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 5 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:836]: Yes. This syntax is only supported in Dataflow runner. `DataflowPackage` has a separate field `name` in addition to `location` and "=" separator allows to prefix `name` to the location of the source e.g. "dataflow.jar=/tmp/foo.jar". I could remove this special syntax but I decided to keep it since it's already exposed to users via `--filesToStage` option so removing it may cause backward compatibility issue.

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 12]
```diff
@@ -1271,6 +1271,11 @@ message DeferredArtifactPayload {
 message ArtifactStagingToRolePayload {
   // A generated staged name (relative path under staging directory).
   string staged_name = 1;
+
+  // (Optional) An artifact name when a runner supports it.
+  // For example, DataflowRunner requires predefined names for some artifacts
+  // such as "dataflow-worker.jar", "windmill_main".
+  string alias_name = 2;
```
[the remark this answers is not in the document]
Comment (ihji) [reply model/pipeline/src/main/proto/beam_runner_api.proto:1278]: This is Dataflow specific requirement. `DataflowPackage` model has two separate fields for `location` and `name`. `staged_name` and `alias_name` correspond to `location` and `name` respectively.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 6 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:329]: removed.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py (hunk 9 above)
Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py:106]: The test tries to download a dev version of apache-beam dependency (which indeed does not exist in pypi) when it constructs the environment.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 8 above)
Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:466]: We need pipeline option to populate artifacts. I think we could either use `from_options` instead and override container image or just leave as is.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 7 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java:448]: I think we can remove those hashing and zipping. It's redundant except test cases. Will update the code.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Run Python PreCommit

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 13]
```diff
@@ -784,7 +877,25 @@ public DataflowPipelineJob run(Pipeline pipeline) {
...
+
+    DataflowPipelineOptions dataflowOptions = options.as(DataflowPipelineOptions.class);
+    String workerHarnessContainerImageURL = DataflowRunner.getContainerImageForJob(dataflowOptions);
+    RunnerApi.Environment defaultEnvironmentForDataflow =
+        Environments.createDockerEnvironment(workerHarnessContainerImageURL);
+
+    sdkComponents.registerEnvironment(
+        defaultEnvironmentForDataflow
+            .toBuilder()
+            .addAllDependencies(getDefaultArtifacts())
```
[the remark this answers is not in the document]
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:894]: It does nothing with the cross-language pipelines. Cross-language dependency should also be a part of the native dependency list here. #11557 improved this so the expansion service could  return its own dependencies.

**Comment by ihji** [comment]
@robertwb Thanks for your comments! PTAL.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 8 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:466]: Let's use from_options.

**Comment by robertwb** [comment]
@ihji Just thinking about timing, were we hoping to get this into 2.22?

**Comment by chamikaramj** [comment]
Yes, I think we need this. Heejong, can you make the jar a blocker for beam 2.22 ?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java [hunk 14]
```diff
@@ -195,7 +187,7 @@ public void testFileWithExtensionPackageNamingAndSize() throws Exception {
     PackageAttributes attr = makePackageAttributes(tmpFile, null);
     DataflowPackage target = attr.getDestination();
 
-    assertThat(target.getName(), RegexMatcher.matches("file-" + HASH_PATTERN + ".txt"));
+    assertThat(target.getName(), RegexMatcher.matches(UUID_PATTERN + ".txt"));
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java:190]: You could just match `".*.txt"` here, rather than hard-coding the uuid format. (Same blow.)

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:861]: OK, we don't have to change this.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java (hunk 14 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java:190]: but it only checks whether the staged file name has the same extension (vs. checks whether the staged file name is in the form of UUID with the same extension)

Comment (robertwb) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java:190]: Is it important, for the purposes of this test (or Dataflow in general) that the staged file name is in the form of a UUID? If not (and I don't think it is) it's better not to test for it. (A test that might be good to add is to see if two same-named files in different directories actually get staged to different places, which is the underlying, important intent.)

**Comment by chamikaramj** [comment]
Retest this please

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java (hunk 14 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java]: Hmm. You're right. UUID is implementation detail behind the uniqueness guarantee. I changed the tests to only check whether it keeps the same extension.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (lukecwik) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java:857]: We also need to make sure we have the capabilities set, I have this PR: https://github.com/apache/beam/pull/11748

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java [hunk 15]
```diff
@@ -53,9 +53,12 @@
 import org.apache.beam.model.pipeline.v1.RunnerApi.DockerPayload;
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:53]: Shouldn't we have a test to show the artifacts were properly set?

**Comment by chamikaramj** [comment]
Heejong, can you please resolve conflicts and push an update to rerun tests ?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java (hunk 15 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslatorTest.java:56]: done.

**Comment by ihji** [comment]
Rebased and fixed conflict. Please rerun the test.

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run Java PreCommit
